lib/OpenQA/Scheduler/Scheduler.pm

```diff
@@ -112,20 +112,32 @@ sub matching_workers {
     return \@filtered;
 }
 
-sub to_be_scheduled {
-    my ($j, $scheduled) = @_;
+sub to_be_scheduled_recurse {
+    my ($j, $scheduled, $taken) = @_;
 
-    my $ci  = $j->{cluster_jobs}->{$j->{id}};
-    my @ret = ($j);
+    return if $taken->{$j->{id}};
+    # if we were called with undef, this is a sign that
+    # the cluster is not fully scheduled (e.g. blocked_by), so
+    # take that as mark but return
+    $taken->{$j->{id}} = $j;
+
+    my $ci = $j->{cluster_jobs}->{$j->{id}};
+    return unless $ci;
     for my $s (@{$ci->{parallel_children}}) {
-        return undef unless $scheduled->{$s};
-        push(@ret, $scheduled->{$s});
+        to_be_scheduled_recurse($scheduled->{$s}, $scheduled, $taken);
     }
     for my $s (@{$ci->{parallel_parents}}) {
-        return undef unless $scheduled->{$s};
-        push(@ret, $scheduled->{$s});
+        to_be_scheduled_recurse($scheduled->{$s}, $scheduled, $taken);
     }
-    return \@ret;
+}
+
+sub to_be_scheduled {
+    my ($j, $scheduled) = @_;
+
+    my %taken;
+    to_be_scheduled_recurse($j, $scheduled, \%taken);
```

Comment:
I have to admit, that now we end up to recurse even more, worries me a bit.

Reply:
we don't recurse more - we just recurse differently. cluster_jobs is recursive too

Reply:
Well, afaics we do it in more points now

Reply:
that's true - feel free to refactor that part. I feel we're duplicating things actually. But sometimes we filter the cluster for scheduled jobs, sometimes for running jobs, sometimes for cloned jobs - the fruit might not hang as low as one things :)